package/cloudshell/cm/customscript/domain/script_downloader.py

```diff
@@ -23,53 +25,100 @@ def __init__(self, logger, cancel_sampler):
+        response = requests.get(url, auth=None, stream=True)
+        response_valid = self._is_response_valid(response, "public")
+
+        if response_valid:
+            file_name = self._get_filename(response)
+
+        # repo is private and token provided
+        if not response_valid and auth.token:
+            self.logger.info("Token provided. Starting download script with Token...")
+            headers = {"Authorization": "Bearer %s" % auth.token }
+            response = requests.get(url, stream=True, headers=headers)
+            
+            response_valid = self._is_response_valid(response, "Token")
+
+            if response_valid:
+                file_name = self._get_filename(response)
+
+        # repo is private and credentials provided, and Token did not provided or did not work. this will NOT work for github. github require Token
+        if not response_valid and (auth.username and auth.password):
+            self.logger.info("username\password provided, Starting download script with username\password...")
+            response = requests.get(url, auth=(auth.username, auth.password) , stream=True)
+            file_name = self._get_filename(response)
+
+            response_valid = self._is_response_valid(response, "username\password")
+
+            if response_valid:
+                file_name = self._get_filename(response)
+
+        if not response_valid:
+            raise Exception('Failed to download script file. please check the logs for more details.')
 
         for chunk in response.iter_content(ScriptDownloader.CHUNK_SIZE):
             if chunk:
                 file_txt += ''.join(chunk)
             self.cancel_sampler.throw_if_canceled()
 
-        self._validate_response(response, file_txt)
+        self._validate_file(file_txt)
 
         return ScriptFile(name=file_name, text=file_txt)
+    
+    def _is_response_valid(self, response, request_method):
+        try:
+            self._validate_response(response)
+            response_valid = True
+        except Exception as ex:
+            failure_message = "failed to Authorize repository with %s" % request_method
+            self.logger.error(failure_message + " :" + str(ex))
+            response_valid = False
 
-    def _validate_response(self, response, content):
-        if response.status_code < 200 or response.status_code > 300:
-            raise Exception('Failed to download script file: '+str(response.status_code)+' '+response.reason+
-                            '. Please make sure the URL is valid, and the credentials are correct and necessary.')
+        return response_valid
 
+    def _validate_file(self, content):
         if content.lstrip('\n\r').lower().startswith('<!doctype html>'):
             raise Exception('Failed to download script file: url points to an html file')
 
+    def _validate_response(self, response):
+        if response.status_code < 200 or response.status_code > 300:            
```

Comment:
we dont expect to get redirect? (sso?) (sometimes 301, 301 are valid response code)

Reply:
this validation is it was before my change. in all my tryouts with github I never got 301. i Don't think we can handle redirect right now since the file build will also have to change (iter_content())